### PR TITLE
Topology Aware Collapsable Graphs by Attributes

### PR DESCRIPTION
**Pull request by silkspace**
g = graphistry.nodes(...).edges(...)

g2 = g.collapse(start_node, attribute='some_attribute', column='some_column_in_nodes_dataframe')

g2.plot() -> produces a graph where all directed nodes with said attribute are collapsed while preserving topology of graph. Nodes without property have edges merged to super node, while all internal nodes collapse. Disconnected super clusters are not merged together, as is the trick with this algorithm. 

Traversal is inefficient for large graphs (scaling is between O(2*(n_edges)*(log(n_edges)) and 2*(n_edges)**(3/2))

### CODE REVIEW

File: graphistry/collapse.py [hunk 1]
```diff
@@ -0,0 +1,532 @@
...
+    """
+    ndf = g._nodes
+    edf = g._edges
+    src = g._source
+    dst = g._destination
+    node = g._node
+    return ndf, edf, src, dst, node
+
+
+def get_children(g, node_id, hops=1):
```
Comment (lmeyerov) [inline graphistry/collapse.py:31]: unncessary as just calling hop?

Comment (silkspace) [reply graphistry/collapse.py:31]: Yes and then I have to set it each time with a dataframe? Less writing?

File: graphistry/collapse.py [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+    ----------------------------------------------------------------------------------
+
+    :param g: graphistry instance
+    :param node_id: `node` to find edges from
+    :param directed: bool, if true, finds all outgoing edges of `node`, default True
+    :param hops: the number of hops from `node` to take, default = 1
+    :returns DataFrame of edges
+    """
+    _, _, src, dst, _ = unpack(g)
+    g2 = get_children(g, node_id, hops=hops)
```
Comment (lmeyerov) [inline graphistry/collapse.py:81]: need to pass direction to get_children/hops?

File: graphistry/collapse.py [hunk 3]
```diff
@@ -0,0 +1,532 @@
...
+    else:
+        if (n2 in edf[edf[src] == n1][dst].values) or (
+            n1 in edf[edf[src] == n2][dst].values
+        ):
+            return True
+    return False
+
+
+def get_edges_of_node(g, node_id, directed=True, hops=1):
+    """
```
Comment (lmeyerov) [inline graphistry/collapse.py:69]: parameter `directed` currently seems to either pick all dst nodes, or all src+dst nodes

while the name `directed` makes it sound like more about the traversal direction (so an arg to `.hop()`) -- clarify the name/comment?

File: graphistry/collapse.py [hunk 4]
```diff
@@ -0,0 +1,532 @@
...
+    :param g: graphistry instance
+    :param node_id: `node` to find edges from
+    :param directed: bool, if true, finds all outgoing edges of `node`, default True
+    :param hops: the number of hops from `node` to take, default = 1
+    :returns DataFrame of edges
+    """
+    _, _, src, dst, _ = unpack(g)
+    g2 = get_children(g, node_id, hops=hops)
+    if directed:
+        edges = g2._edges[dst]
```
Comment (lmeyerov) [inline graphistry/collapse.py:83]: need a `.drop_duplicates()` here too... move to be part of the return?

File: graphistry/collapse.py [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+#
+#
+# def remove_edges(g, edgelist):
+#     # src_edges, dst_edges = np.array(edgelist).T
+#     ndf, edf, src, dst, node = unpack(g)
+#     tedf = edf[~edf[src].isin(edgelist) & ~edf[dst].isin(edgelist)]
+#     return tedf
+
+
+def get_edges_in_out_cluster(g, node_id, attribute, column, directed=True):
```
Comment (lmeyerov) [inline graphistry/collapse.py:105]: mypy types

File: graphistry/collapse.py [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+    -----------------------------------
+
+    :param name: node name
+    :returns wrapped node name
+    """
+    name = str(name)
+    if WRAP in name:  # idempotency
+        return name
+    return f"{WRAP}{name}{WRAP}"
+
```
Comment (lmeyerov) [inline graphistry/collapse.py:209]: so if `"~"` is in a node id, what happens?

File: graphistry/collapse.py (hunk 3 above)
Comment (silkspace) [reply graphistry/collapse.py:69]: Yes, agreed

File: graphistry/collapse.py (hunk 4 above)
Comment (silkspace) [reply graphistry/collapse.py:83]: 4 eyes are better than 2, thanks

File: graphistry/collapse.py (hunk 1 above)
Comment (lmeyerov) [reply graphistry/collapse.py:31]: hmm maybe we can make `hop` signature more convenient then:

ex:  this could be replaced with an even short `g.hop([node_id], hops)`, where we teach `hop()` that if instead of nodes=pd.DataFrame, it gets `List[str]`, it auto-coerces into a nodes df for them

Comment (lmeyerov) [reply graphistry/collapse.py:31]: so `hop(self, nodes=Union[ pd.DataFrame, List[str] ], ...`

File: graphistry/collapse.py (hunk 6 above)
Comment (silkspace) [reply graphistry/collapse.py:209]: The age old problem of escaping/split/replace, open to suggestions -- might even have WRAP be parameter

`g.collapse(..., escaping_token='~')` 

?

Comment (lmeyerov) [reply graphistry/collapse.py:209]: (that's ok, but maybe we can detect-and-throw at the beginning if that's the case?)

File: graphistry/collapse.py [hunk 7]
```diff
@@ -0,0 +1,532 @@
...
+
+def melt(ndf, node) -> str:
+    """
+        Reduces node if in cluster store, otherwise passes it through.
+    ex:
+        node = "4" will take any sequence from get_cluster_store_keys, "1 2 3", "4 3" and returns "1 2 3 4"
```
Comment (lmeyerov) [inline graphistry/collapse.py:215]: `node = "4"`... or `node = "3"` ? (trying to follow)

File: graphistry/collapse.py [hunk 8]
```diff
@@ -0,0 +1,532 @@
...
+    # THIS IS IMPORTANT FUNCTION -- it is where we wrap the parent/child in WRAP
+    # if child in cluster group, we melt it
+    ckey = in_cluster_store_keys(ndf, child)
+
+    if ckey:
+        new_parent_name = melt(ndf, child)
+    else:  # if not, then append child to parent as the start of a new cluster group
+        # might have to escape parent and child if node names are dumb eg, 'this value key'
+        new_parent_name = f"{wrap_key(parent)} {wrap_key(child)}"
+    if VERBOSE:
```
Comment (lmeyerov) [inline graphistry/collapse.py:260]: `logger.debug(...)`

File: graphistry/collapse.py [hunk 9]
```diff
@@ -0,0 +1,532 @@
...
+    :param g: graphistry instance
+    :param parent: `node` with `attribute` in `column`
+    :param child: `node` with `attribute` in `column`
+    :returns: graphistry instance, new_parent_name of super cluster
+    """
+
+    ndf, edf, src, dst, node = unpack(g)
+
+    new_parent_name = get_new_node_name(ndf, parent, child)
+
```
Comment (lmeyerov) [inline graphistry/collapse.py:285]: hmmm this is a mutation... probably better not to return anything

File: graphistry/collapse.py [hunk 10]
```diff
@@ -0,0 +1,532 @@
...
+    for node in incluster:
+        edf.loc[edf[src] == parent, COLLAPSE_SRC] = new_parent_name
+        edf.loc[edf[dst] == parent, COLLAPSE_DST] = new_parent_name
+
+        edf.loc[edf[dst] == node, COLLAPSE_DST] = new_parent_name
+        edf.loc[edf[src] == node, COLLAPSE_SRC] = new_parent_name
+
+    g._edges = edf
+    return g
```
Comment (lmeyerov) [inline graphistry/collapse.py:320]: again, this is really a mutation of existing g, not returning a new g, so better to not return anything

File: graphistry/collapse.py [hunk 11]
```diff
@@ -0,0 +1,532 @@
...
+    -------------------------------------------------------------------------
+
+    :param g: graphistry instance
+    :returns graphistry instance
+    """
+    ndf, edf, src, dst, node = unpack(g)
+    if COLLAPSE_NODE not in ndf.columns:
+        ndf[COLLAPSE_NODE] = DEFAULT_VAL
+        ndf[node] = ndf[node].astype(str)
+        print(f'Converted ndf to type({type(ndf[node].values[0])})')
```
Comment (lmeyerov) [inline graphistry/collapse.py:350]: logger.debug

File: graphistry/collapse.py [hunk 12]
```diff
@@ -0,0 +1,532 @@
...
+        g._nodes = ndf
+    if COLLAPSE_SRC not in edf.columns:
+        edf[COLLAPSE_SRC] = DEFAULT_VAL
+        edf[COLLAPSE_DST] = DEFAULT_VAL
+        edf[src] = edf[src].astype(str)
+        edf[dst] = edf[dst].astype(str)
+        print(f'Converted edf to type({type(edf[src].values[0])})')
```
Comment (lmeyerov) [inline graphistry/collapse.py:357]: logger.debug

File: graphistry/collapse.py [hunk 13]
```diff
@@ -0,0 +1,532 @@
...
+def check_default_columns_present_and_coerce_to_string(g):
+    """
+        Helper to set COLLAPSE columns to nodes and edges dataframe, while converting src, dst, node to dtype(str)
+
+    -------------------------------------------------------------------------
+
+    :param g: graphistry instance
+    :returns graphistry instance
+    """
+    ndf, edf, src, dst, node = unpack(g)
```
Comment (lmeyerov) [inline graphistry/collapse.py:346]: this is a mutation on input ndf, probably should be:

```python
ndf2 = ndf.copy(deep=False) # or True if later you mutate the cells.. which you seem to...
...
g2 = g.nodes(ndf2)
...
return g2 
```

File: graphistry/collapse.py [hunk 14]
```diff
@@ -0,0 +1,532 @@
...
+        if has_property(g, parent, attribute, column):  # if (T, *)
+            # add start node to super node index
+            tkey = f"{parent} {parent}"  # it will reduce this to `parent` but can add to `seen`
+            if tkey not in compute_key:  # its love!
+                seen[tkey] = 1
+                g, _ = collapse_nodes(g, parent, parent)
+            if has_property(g, child, attribute, column):  # if (T, T)
+                # add to seen
+                seen[compute_key] = 1
+                if VERBOSE:
```
Comment (lmeyerov) [inline graphistry/collapse.py:409]: logger.debug

Comment (lmeyerov) [reply graphistry/collapse.py:409]: and `if logger.isEnabledFor(logging.DEBUG):  ...`

File: graphistry/collapse.py [hunk 15]
```diff
@@ -0,0 +1,532 @@
...
+
+    :param g: graphistry instance
+    :param child: child node to start traversal, for first traversal, set child=parent or vice versa.
+    :param parent: parent node to start traversal, in main call, this is set to child.
+    :param attribute: attribute to collapse by
+    :param column: column in nodes dataframe to collapse over.
+    :returns graphistry instance with collapsed nodes.
+
+    """
+    g = check_default_columns_present_and_coerce_to_string(g)
```
Comment (lmeyerov) [inline graphistry/collapse.py:390]: this gets called on every recursion, is that right? or just should be at entrypoint?

File: graphistry/collapse.py [hunk 16]
```diff
@@ -0,0 +1,532 @@
...
+    --------------------------------------------------------------------------------------------------------------------
+
+    :param g: graphistry instance
+    :param self_edges: bool, whether to keep duplicates from ndf, edf, default False
+    :param unwrap: bool, whether to unwrap node text with `~`, default True
+    :param remove_collapse: bool, whether to drop COLLAPSE columns from dataframe, default True
+    :returns final graphistry instance
+    """
+    # at the end of collapse, move anything untouched to new graph
+    ndf, edf, src, dst, node = unpack(g)
```
Comment (lmeyerov) [inline graphistry/collapse.py:455]: these are mutations on ndf/edge, should probably be a `ndf.copy(deep=True)`

File: graphistry/collapse.py [hunk 17]
```diff
@@ -0,0 +1,532 @@
...
+    from time import time
+
+    __doc__ = collapse.__doc__
+    
+    
+    
+    n_edges = len(g._edges)
+    complexity_min = int(2 * n_edges * np.log(n_edges))
+    complexity_max = int(2 * n_edges ** (3 / 2))
+    if VERBOSE:
```
Comment (lmeyerov) [inline graphistry/collapse.py:509]: logger.debug

File: graphistry/collapse.py [hunk 18]
```diff
@@ -0,0 +1,532 @@
...
+        ndf = ndf.drop(columns=[COLLAPSE_NODE])
+        edf = edf.drop(columns=[COLLAPSE_SRC, COLLAPSE_DST])
+
+    # set the dataframes
+    g._nodes = ndf
+    g._edges = edf
+    return g
+
+
+def collapse_by(g, parent, start_node, attribute, column, seen):
```
Comment (lmeyerov) [inline graphistry/collapse.py:488]: do you have a sense of how many ~seconds this is running on say:

- 100 node / 1K edges
- 1K nodes / 10K edges
- 10K nodes / 100K edges

**Review by lmeyerov — CHANGES_REQUESTED** [review]
Main changes:
- add mypy types
- print -> logger.debug for logger of name __name__
- for mutating funcs parents, `ndf2 = ndf.copy(deep=True); return g.nodes(ndf2)`
- for mutating funcs helpers, do not return `g`


Ideally, improve `hop()` sig instead of adding a new alias

Am curious about the perf on diff size graphs. OK if not 100K node grpahs, but would hope we can do 10K node (100K edge) ones. Not sure though, `hop()` etc may be slow as was meant for bulk calls, not individual.

File: graphistry/collapse.py (hunk 9 above)
Comment (silkspace) [reply graphistry/collapse.py:288]: Fixed

File: graphistry/collapse.py (hunk 10 above)
Comment (silkspace) [reply graphistry/collapse.py:300]: Fixed

File: graphistry/collapse.py (hunk 13 above)
Comment (silkspace) [reply graphistry/collapse.py:330]: I handle this in outer scope, in compute.py

`        from .collapse import collapse_by

        #__doc__ += collapse_by.__doc__
        g = self.bind()
        # TODO FIXME CHECK SELF LOOPS?
        return collapse_by(
            g,
            start_node=node,
            parent=node,
            attribute=attribute,
            column=column,
            seen={},
            self_edges=self_edges,
            unwrap=unwrap,
            remove_collapse=remove_collapse,
        )
`

File: graphistry/collapse.py (hunk 15 above)
Comment (silkspace) [reply graphistry/collapse.py:381]: Every recursion, but lightweight bools overall. I can move it out of scope, and change `collapse` -> `_collapse`

File: graphistry/collapse.py (hunk 18 above)
Comment (silkspace) [reply graphistry/collapse.py]: Get to this tomorrow in tests

File: graphistry/collapse.py (hunk 13 above)
Comment (lmeyerov) [reply graphistry/collapse.py:330]: @silkspace  `g = self.bind()` is a shallow copy: it'll do `res._nodes = self._nodes` instead of `res._nodes = self._nodes.copy()` . Likewise, I'm fuzzy on pandas semantics around mutation, so you may even need `nodes.copy(deep=True)`.

Comment (lmeyerov) [reply graphistry/collapse.py:330]: a test can be:

```python
g = graphistry.edges(...)
g2 = g.bind()
g2._edges[...] = something # mutate...
print(g._edges)
```

File: graphistry/Plottable.py [hunk 19]
```diff
@@ -87,6 +87,18 @@ def get_topological_levels(
         raise RuntimeError('should not happen')
         return self
 
+    def collapse(
```
Comment (lmeyerov) [inline graphistry/Plottable.py:90]: Is there some sort of outer loop that is supposed to be used in a typical case? I'd think we'd want 2 methods:

1. clarify via renaming the current entrypoint as `collapse_adjacent_by_col_val_match(g, root, col, val)`
  - later, we cab generalize to the still-vectorizable form `collapse_adjacent_by_dict_match(g, root, {col1: val1, col2: val2, ...})`
2. add typical outer loop `collapse_adjacent_by_col_match(g, root, col)`
  - basically: `for v in g._nodes[col].unique(): g = collapse(g, root, col, v)`
  - later do the same generalization

We can land without for now..

**Comment by lmeyerov** [comment]
Merging!

I'll file follow-up tickets:
- generalize interface: pass in a kv-dict instead of kv-pair, and pass in a col list instead of kv-pair
- reductions, including provenance tracking
- optimization
